src/main/scala/com/scality/clueso/MetadataIngestionPipeline.scala

```diff
@@ -0,0 +1,187 @@
+          val _locationArray =  locationArray.asInstanceOf[ArrayNode]
+          (1 until locationArray.size()).foreach(i => _locationArray.remove(i))
+          valueNode.replace("location", _locationArray)
+        }
+      }
+    }
+
+    val userMd = if (rootNode.has("userMd")) {
+      valueNode.path("userMd").asInstanceOf[ObjectNode]
+    } else {
+      valueNode.putObject("userMd")
+    }
+
+    for (fieldName <- metadataFieldNames) {
+      userMd.set(fieldName, valueNode.path(fieldName).deepCopy().asInstanceOf[JsonNode])
+      valueNode.remove(fieldName)
+    }
+
+    valueNode.replace("userMd", userMd)
+    rootNode.replace("value", valueNode)
+
+    mapper.writeValueAsString(rootNode)
+  }
+}
+
+object EventMessageRewriterWrapper {
+  val deser = new EventMessageRewriter
+}
+
+object MetadataIngestionPipeline extends LazyLogging {
+
+  val msgRewriteFun = (bytes: Array[Byte]) =>
+    EventMessageRewriterWrapper.deser.rewriteMsg(bytes)
+
+  val maxOpIndexFun = (compactionRecordInterval:Long, value : String) => {
+    val recordNo = value.substring(0, 12).toLong
+
+    if (recordNo % compactionRecordInterval == 0) {
+      recordNo
+    } else {
+      recordNo + compactionRecordInterval - recordNo % compactionRecordInterval
+    }
+  }
+
+  import org.apache.spark.sql.functions.udf
+  val msg_rewrite = udf(msgRewriteFun)
+  val max_op_index = udf(maxOpIndexFun)
+
+  /**
+    * Applies projections and conditions to incoming data frame
+    * Ignores malformed json and garbage payloads
+    * Ignores events from specified bucket name
+    *
+    * @param bucketNameToFilterOut
+    * @param eventStream
+    * @return
+    */
+  def filterAndParseEvents(bucketNameToFilterOut : String, eventStream: DataFrame)(implicit spark : SparkSession, config: CluesoConfig) = {
+
+    var df = eventStream.select(
+      col("timestamp").cast(TimestampType).as("kafkaTimestamp"),
+      msg_rewrite(col("value")).as("content")
+    )
+
+    // defensive filtering to not process kafka garbage
+    df = df.filter(col("content").isNotNull)
+          .filter(length(col("content")).gt(3))
+          .select(
+            col("kafkaTimestamp"),
+            from_json(col("content"), CluesoConstants.eventSchema).alias("event")
+          )
+
+    df = df.filter(col("event").isNotNull.and(col("event.type").isNotNull))
+      .withColumn("key", when(
+        col("event").isNotNull.and(
+          col("event.key").isNotNull
+        ), col("event.key")).otherwise("")
+      )
+      .withColumn("bucket",
+        when(
+          col("event").isNotNull.and(
+            col("event.bucket").isNotNull
+          ), col("event.bucket")).otherwise("NOBUCKET")
+      )
+      .withColumn("type", col("event.type"))
+      .withColumn("opIndex", col("event.opIndex"))
+      .withColumn("maxOpIndex", max_op_index(lit(config.compactionRecordInterval), col("event.opIndex")))
```

Comment:
why do you have to use lit here?

Reply:
lit means literal. I want the Spark Plan to call that user defined function, with the value of that variable (which is only present in the driver program) as a literal. By doing that it will 'lock' the current value into the spark plan